test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py

```diff
@@ -0,0 +1,153 @@
+# Owner(s): ["oncall: distributed"]
+
+import functools
+import itertools
+import sys
+import torch
+
+from torch import distributed as dist
+from torch.distributed.fsdp.fully_sharded_data_parallel import ShardingStrategy
+from torch.distributed.fsdp import MixedPrecision, CPUOffload
+from torch.distributed.fsdp.sharded_grad_scaler import ShardedGradScaler
+from torch.testing._internal.common_fsdp import DummyProcessGroup, subtest_name, FSDPInitMode, NestedWrappedModule, FSDPTest
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import (
+    TestCase, run_tests,
+    instantiate_parametrized_tests,
+    parametrize,
+    TEST_WITH_DEV_DBG_ASAN,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+params = "cpu_offload,sharding_strategy,mixed_precision"
+cpu_offload_config = [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
+sharding_strategy_config = [ShardingStrategy.SHARD_GRAD_OP, None]
+mixed_precision = ["enable_mixed_precision", None]
+
+configs = list(itertools.product(cpu_offload_config,
+                                 sharding_strategy_config,
+                                 mixed_precision))
+test_name_mapping = {
+    str(CPUOffload(offload_params=True)): "offload_true",
+    str(CPUOffload(offload_params=False)): "offload_false",
+    str(ShardingStrategy.SHARD_GRAD_OP): "shard_grad_op",
+    "enable_mixed_precision": "mixed_precision"
+}
+
+subtest_name = functools.partial(subtest_name, test_name_mapping)
+
+
+class TestShardGradScaler(TestCase):
```

Comment:
These tests don't really seem to depend on shards / multiple ranks? Should we file an issue to test this against traditional grad scaler and verify we have parity?

Reply:
They dont directly depend on shard/multiple ranks. GradScaler.py has a check where if cuda is not available, it disables GradScaling. https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L116